README.md
# PictureInPicture

[![Platform](https://img.shields.io/cocoapods/p/PictureInPicture.svg?style=flat)](http://cocoapods.org/pods/PictureInPicture)
![Language](https://img.shields.io/badge/language-Swift%203.1-orange.svg)
[![CocoaPods](https://img.shields.io/cocoapods/v/PictureInPicture.svg?style=flat)](http://cocoapods.org/pods/PictureInPicture)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)
![License](https://img.shields.io/github/license/malt03/PictureInPicture.svg?style=flat)

![ScreenShot](https://raw.githubusercontent.com/malt03/PictureInPicture/master/README/Screenshot.gif)

## Usage
### Configure
If you want to change from default value.

```swift
let shadowConfig = PictureInPicture.ShadowConfig(color: .black, offset: .zero, radius: 10, opacity: 1)
PictureInPicture.configure(movable: true,
                           scale: 0.3,
                           margin: 10,
                           defaultEdge: .left,
                           shadowConfig: shadowConfig)
```

#### Default Config
```swift
PictureInPicture.configure(movable: true,
                           scale: 0.2,
                           margin: 8,
                           defaultEdge: .right,
                           shadowConfig: .default)
```

#### Default Shadow Config
```swift
ShadowConfig(color: .black, offset: .zero, radius: 5, opacity: 0.5)
```

### Main Functions
```swift
PictureInPicture.shared.present(with: viewController) // Present
PictureInPicture.shared.dismiss()                     // Dismiss
PictureInPicture.shared.makeSmaller()                 // Make Smaller
PictureInPicture.shared.makeLarger()                  // Make Larger
PictureInPicture.shared.presentedViewController       // Get presented ViewController
```

### Notifications
- PictureInPictureMadeSmaller
- PictureInPictureMadeLarger
- PictureInPictureDidBeginMakingSmaller
- PictureInPictureDidBeginMakingLarger
- PictureInPictureMoved
  - PictureInPictureOldCornerUserInfoKey # PictureInPicture.Corner
  - PictureInPictureNewCornerUserInfoKey # PictureInPicture.Corner
- PictureInPictureDismissed

### Other Constants
- UIWindowLevelPictureInPicture

## Installation

### Via CocoaPods
```ruby
pod "PictureInPicture"
```

### Via Carthage
```ruby
github "malt03/PictureInPicture"
```

## Author

Koji Murata, malt.koji@gmail.com

## License

PictureInPicture is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Change Log

## [0.2.4](https://github.com/malt03/PictureInPicture/tree/0.2.4) (2017-08-08)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.2.3...0.2.4)

## [0.2.3](https://github.com/malt03/PictureInPicture/tree/0.2.3) (2017-08-08)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.2.2...0.2.3)

## [0.2.2](https://github.com/malt03/PictureInPicture/tree/0.2.2) (2017-07-28)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.2.1...0.2.2)

## [0.2.1](https://github.com/malt03/PictureInPicture/tree/0.2.1) (2017-07-28)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.2.0...0.2.1)

**Closed issues:**

- I've been looking for a library like this but for objective C [\#1](https://github.com/malt03/PictureInPicture/issues/1)

## [0.2.0](https://github.com/malt03/PictureInPicture/tree/0.2.0) (2017-07-27)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.1.14...0.2.0)

## [0.1.14](https://github.com/malt03/PictureInPicture/tree/0.1.14) (2017-07-26)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.1.13...0.1.14)

## [0.1.13](https://github.com/malt03/PictureInPicture/tree/0.1.13) (2017-07-26)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.1.12...0.1.13)

## [0.1.12](https://github.com/malt03/PictureInPicture/tree/0.1.12) (2017-07-25)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.1.11...0.1.12)

## [0.1.11](https://github.com/malt03/PictureInPicture/tree/0.1.11) (2017-07-25)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.1.10...0.1.11)

## [0.1.10](https://github.com/malt03/PictureInPicture/tree/0.1.10) (2017-07-25)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.1.9...0.1.10)

## [0.1.9](https://github.com/malt03/PictureInPicture/tree/0.1.9) (2017-07-20)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.1.8...0.1.9)

## [0.1.8](https://github.com/malt03/PictureInPicture/tree/0.1.8) (2017-07-20)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.1.7...0.1.8)

## [0.1.7](https://github.com/malt03/PictureInPicture/tree/0.1.7) (2017-07-19)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.1.6...0.1.7)

## [0.1.6](https://github.com/malt03/PictureInPicture/tree/0.1.6) (2017-07-19)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.1.5...0.1.6)

## [0.1.5](https://github.com/malt03/PictureInPicture/tree/0.1.5) (2017-07-19)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.1.4...0.1.5)

## [0.1.4](https://github.com/malt03/PictureInPicture/tree/0.1.4) (2017-07-19)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.1.3...0.1.4)

## [0.1.3](https://github.com/malt03/PictureInPicture/tree/0.1.3) (2017-07-19)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.1.2...0.1.3)

## [0.1.2](https://github.com/malt03/PictureInPicture/tree/0.1.2) (2017-07-19)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.1.1...0.1.2)

## [0.1.1](https://github.com/malt03/PictureInPicture/tree/0.1.1) (2017-06-16)
[Full Changelog](https://github.com/malt03/PictureInPicture/compare/0.1.0...0.1.1)

## [0.1.0](https://github.com/malt03/PictureInPicture/tree/0.1.0) (2017-06-16)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

PictureInPicture/Classes/PictureInPicture.swift
//
//  PictureInPicture.swift
//  Pods
//
//  Created by Koji Murata on 2017/07/17.
//
//

import Foundation

extension Notification.Name {
  public static let PictureInPictureMadeSmaller = Notification.Name(rawValue: "PictureInPictureMadeSmaller")
  public static let PictureInPictureMadeLarger = Notification.Name(rawValue: "PictureInPictureMadeLarger")
  public static let PictureInPictureDidBeginMakingSmaller = Notification.Name(rawValue: "PictureInPictureDidBeginMakingSmaller")
  public static let PictureInPictureDidBeginMakingLarger = Notification.Name(rawValue: "PictureInPictureDidBeginMakingLarger")
  public static let PictureInPictureMoved = Notification.Name(rawValue: "PictureInPictureMoved")
  public static let PictureInPictureDismissed = Notification.Name(rawValue: "PictureInPictureDismissed")
}

public let PictureInPictureOldCornerUserInfoKey = "PictureInPictureOldCornerUserInfoKey"
public let PictureInPictureNewCornerUserInfoKey = "PictureInPictureNewCornerUserInfoKey"
public let UIWindowLevelPictureInPicture = UIWindowLevelNormal + 1

public final class PictureInPicture {
  public struct ShadowConfig {
    let color: UIColor
    let offset: CGSize
    let radius: CGFloat
    let opacity: Float
    
    public init(color: UIColor = .black, offset: CGSize = .zero, radius: CGFloat = 5, opacity: Float = 0.5) {
      self.color = color
      self.offset = offset
      self.radius = radius
      self.opacity = opacity
    }
    
    public static var `default`: ShadowConfig {
      return ShadowConfig()
    }
  }
  
  public static func configure(movable: Bool = true,
                               scale: CGFloat = 0.2,
                               margin: CGFloat = 8,
                               defaultEdge: HorizontalEdge = .right,
                               shadowConfig: ShadowConfig = .default) {
    self.movable = movable
    self.scale = scale
    self.margin = margin
    self.defaultEdge = defaultEdge
    self.shadowConfig = shadowConfig
  }
  
  private(set) static var movable = true
  private(set) static var scale = CGFloat(0.2)
  private(set) static var margin = CGFloat(8)
  private(set) static var shadowConfig = ShadowConfig.default
  private static var defaultEdge = HorizontalEdge.right
  
  static var defaultCorner: Corner {
    return Corner(.bottom, defaultEdge)
  }
  
  public static let shared = PictureInPicture()
  
  public func present(with viewController: UIViewController, makeLargerIfNeeded: Bool = true) {
    windowCreateIfNeeded.present(with: viewController, makeLargerIfNeeded: makeLargerIfNeeded)
  }
  
  public func makeLarger() {
    window?.applyLarge()
  }
  
  public func makeSmaller() {
    window?.applySmall()
  }
  
  public func dismiss(animation: Bool = true) {
    window?.dismiss(animation: animation)
    window = nil
  }
  
  public var presentedViewController: UIViewController? {
    return window?.rootViewController
  }
  
  public var currentCorner: Corner {
    return window?.currentCorner ?? PictureInPicture.defaultCorner
  }
  
  private init() {
    prepareNotification()
  }
  
  private var windowCreateIfNeeded: PictureInPictureWindow {
    if let w = window { return w }
    let w = PictureInPictureWindow {
      self.keyWindow?.makeKeyAndVisible()
      self.window = nil
    }
    window = w
    return w
  }
  private(set) var window: PictureInPictureWindow?
  private var keyWindow = UIApplication.shared.keyWindow
  
  private func prepareNotification() {
    NotificationCenter.default.addObserver(self, selector: #selector(windowDidBecomeKey), name: .UIWindowDidBecomeKey, object: nil)
    NotificationCenter.default.addObserver(self, selector: #selector(madeLarger), name: .PictureInPictureMadeLarger, object: nil)
    NotificationCenter.default.addObserver(self, selector: #selector(didBeginMakingSmaller), name: .PictureInPictureDidBeginMakingSmaller, object: nil)
  }
  
  @objc private func windowDidBecomeKey() {
    if window == UIApplication.shared.keyWindow { return }
    keyWindow = UIApplication.shared.keyWindow
  }
  
  @objc private func madeLarger() {
    window?.makeKeyAndVisible()
  }
  
  @objc private func didBeginMakingSmaller() {
    keyWindow?.makeKeyAndVisible()
  }
  
  public enum HorizontalEdge {
    case left
    case right
  }
  
  public enum VerticalEdge {
    case top
    case bottom
  }
  
  public enum Corner {
    case topLeft
    case topRight
    case bottomLeft
    case bottomRight
    
    public var verticalEdge: VerticalEdge {
      switch self {
      case .topLeft, .topRight: return .top
      case .bottomLeft, .bottomRight: return .bottom
      }
    }
    
    public var horizontalEdge: HorizontalEdge {
      switch self {
      case .topLeft, .bottomLeft: return .left
      case .topRight, .bottomRight: return .right
      }
    }
    
    public init(_ verticalEdge: VerticalEdge, _ horizontalEdge: HorizontalEdge) {
      switch verticalEdge {
      case .top:
        switch horizontalEdge {
        case .left:  self = .topLeft
        case .right: self = .topRight
        }
      case .bottom:
        switch horizontalEdge {
        case .left:  self = .bottomLeft
        case .right: self = .bottomRight
        }
      }
    }
  }
}
